planner/core/rule_constant_propagation.go

```diff
@@ -0,0 +1,284 @@
+	case RightOuterJoin:
+		getConstantPredicateFromRight = true
+	case InnerJoin:
+		getConstantPredicateFromLeft = true
+		getConstantPredicateFromRight = true
+	default:
+		return
+	}
+	var candidateConstantPredicates []expression.Expression
+	if getConstantPredicateFromLeft {
+		candidateConstantPredicates = logicalJoin.children[0].recursiveFindAndPullUpConstantPredicates()
+	}
+	if getConstantPredicateFromRight {
+		candidateConstantPredicates = append(candidateConstantPredicates, logicalJoin.children[1].recursiveFindAndPullUpConstantPredicates()...)
+	}
+	if len(candidateConstantPredicates) == 0 {
+		return
+	}
+
+	// step2: add selection above of LogicalJoin
+	return addCandidateSelection(logicalJoin, currentChildIdx, parentPlan, candidateConstantPredicates, opt)
+}
+
+func (*baseLogicalPlan) recursiveFindAndPullUpConstantPredicates() []expression.Expression {
+	// Only LogicalProjection and LogicalSelection can get constant predicates
+	// Other Logical plan return nil
+	return nil
+}
+
+func (selection *LogicalSelection) recursiveFindAndPullUpConstantPredicates() []expression.Expression {
+	var result []expression.Expression
+	for _, candidatePredicate := range selection.Conditions {
+		// the candidate predicate should be a constant and compare predicate
+		match := validCompareConstantPredicate(candidatePredicate)
+		if match {
+			result = append(result, candidatePredicate)
+		}
+	}
+	return result
+}
+
+func (projection *LogicalProjection) recursiveFindAndPullUpConstantPredicates() []expression.Expression {
+	// projection has no column expr
+	if !canProjectionBeEliminatedLoose(projection) {
+		return nil
+	}
+	candidateConstantPredicates := projection.children[0].recursiveFindAndPullUpConstantPredicates()
+	// replace predicate by projection expr
+	// candidate predicate : a=1
+	// projection: a as a'
+	// result predicate : a'=1
+	replace := make(map[string]*expression.Column)
+	for i, expr := range projection.Exprs {
+		replace[string(expr.HashCode(nil))] = projection.Schema().Columns[i]
+	}
+	result := make([]expression.Expression, 0, len(candidateConstantPredicates))
+	for _, predicate := range candidateConstantPredicates {
+		// The column of predicate must exist in projection exprs
+		columns := expression.ExtractColumns(predicate)
+		// The number of columns in candidate predicate must be 1.
+		if len(columns) != 1 {
+			continue
+		}
+		if replace[string(columns[0].HashCode(nil))] == nil {
+			// The column of predicate will not appear on the upper level
+			// This means that this predicate does not apply to the constant propagation optimization rule
+			// For example: select * from t, (select b from s where s.a=1) tmp where t.b=s.b
+			continue
+		}
+		clonePredicate := predicate.Clone()
+		ResolveExprAndReplace(clonePredicate, replace)
+		result = append(result, clonePredicate)
+	}
+	return result
+}
+
+// validComparePredicate checks if the predicate is an expression like [column '>'|'>='|'<'|'<='|'=' constant].
```

Comment:
Should we support "IN" as well?

Reply:
IN I feel it is also possible. However, the customer does not use IN currently. I wonder if we can add it when a customer needs it.